### ISSUE TITLE
Key derivation rounds and saving/closing delay

### ISSUE THREAD

**Issue by tflo** [mstarke/MacPass#658]
Not an issue, just a question out of curiosity:

With the key derivation rounds set to, let's say, 50000000 I get a 2-seconds delay when…

- opening the database
- saving the database
- closing the database

However, the same database with KeePassXC produces the 2-seconds delay only when opening the database; not when saving, not when closing.

So, in layman's terms, it seems that both apps have to go through the derivation rounds when *de*crypting (which of course is the whole point of the thing), but only MacPass goes through the derivation rounds also when *en*crypting.

Why is this so?

– Tom

**Comment by mstarke** [comment]
KeePassXC stores the transformed key after it was computed once. This happens on load and if you do not change the key or the number of rounds the stored data is used for writing. KeePass for windows as well as MacPass does not reuse any seeds but regernerates them every time the key is derived (which happens on every save). I have no clue why KeePassX opts to not do this since it weakens the database agains attacks.

**Comment by mstarke** [comment]
I just opened an issue[1] on KeePassXC to see what the develpers think about this behaviour.

[1] https://github.com/keepassxreboot/keepassxc/issues/1051

**Comment by tflo** [comment]
Thanks for the explanation.

I've seen the guy on the [keepassxc repo](https://github.com/keepassxreboot/keepassxc/issues/1051#event-1283788174) has tagged it as high priority bug. It seems my question was not as naive as I thought, just posted on the wrong repo ;)

> I have no clue why KeePassXC opts to not do this _since it weakens the database against attacks_.

Do you mean the KeePassXC database is weakened while open, or always, also when closed?

– Tom

**Comment by mstarke** [comment]
It's a good catch :-)

The weakness is introduced by not changing the random seed. Only the master seed is generated all the time but not the seed for the key derivation and thus it reduces the amount of data an attacker has to guess significantly. I have no numbers on how much weaker this is. But even Dominik mentioned it on the KeePas website: https://keepass.info/help/base/security.html#secdictprotect

**Comment by mstarke** [comment]
I'm closing this since I guess everything is clear now.
